spec/datadog/core/utils/network_spec.rb

```diff
@@ -0,0 +1,127 @@
+require 'spec_helper'
+
+require 'datadog/core/utils/network'
+
+RSpec.describe Datadog::Core::Utils::Network do
+  describe '.raw_ip_from_request_headers' do
+    context 'with default IP headers' do
+      context 'iterates over the deafult headers in order' do
+        it 'returns the first valid public IP value' do
+          headers = Datadog::Core::HeaderCollection.from_hash(
+            { 'X-Forwarded-For' => '10.42.42.42',
+              'True-Client-Ip' => '43.43.43.43',
+              'X-Cluster-Client-Ip' => '10.0.0.1', }
+          )
+
+          result = described_class.raw_ip_from_request_headers(headers)
+          expect(result).to eq('43.43.43.43')
+        end
+      end
+
+      context 'multiple IP addresses present in the header' do
+        it 'returns the first valid public IP address' do
+          headers = Datadog::Core::HeaderCollection.from_hash({ 'X-Forwarded-For' => '10.42.42.42,43.43.43.43,fe80::1' })
+
+          result = described_class.raw_ip_from_request_headers(headers)
+          expect(result).to eq('43.43.43.43')
+        end
+      end
+
+      context 'with custom header value' do
+        it 'returns the IP value if valid public address' do
+          headers = Datadog::Core::HeaderCollection.from_hash(
+            {
+              'X-Forwarded-For' => '64.233.161.147',
+              'test-header' => '43.43.43.43',
+            }
+          )
+
+          result = described_class.raw_ip_from_request_headers(headers, ['test-header'])
+          expect(result).to eq('43.43.43.43')
+        end
+
+        it 'returns nil if header not present' do
+          headers = Datadog::Core::HeaderCollection.from_hash({})
+
+          result = described_class.raw_ip_from_request_headers(headers, ['test-header'])
+          expect(result).to be_nil
+        end
+
+        it 'returns nil if header value is not valid' do
+          headers = Datadog::Core::HeaderCollection.from_hash({ 'test-header' => 'dd' })
+
+          result = described_class.raw_ip_from_request_headers(headers, ['test-header'])
+          expect(result).to be_nil
+        end
+      end
+
+      it 'returns nil if no public valid IP addresss present in the headers' do
+        headers = Datadog::Core::HeaderCollection.from_hash(
+          { 'X-Forwarded-For' => '10.42.42.42' }
+        )
+
+        result = described_class.raw_ip_from_request_headers(headers)
+        expect(result).to be_nil
+      end
+    end
+  end
+
+  describe '.parse_ip' do
+    context 'valid IP' do
+      it 'returns IPAddr instance' do
```

Comment:
The `valid_ip?` is different than `parsed_ip` on the `parsed_ip` we strip port and zone specifier (IPV6) before trying to parse it to an `IPAddr`

So some of the previous test would not fail, that is way the list got reduce

Reply:
> So some of the previous test would not fail, that is way the list got reduce

Does that mean that we allow more ips, or disallow more? Since I asked about both, I'm not quite sure which you were answering about (or both?).

Reply:
> Does that mean that we allow more ips  , or disallow more?

I do not think so, it is the same amount, but the logic between `valid_ip` and `parsed_ip` is different, so the previous test cases would not result in the same outcome. 

In the previous code, we always called `valid_ip` after stripping the ip from the port and zone. 

```ruby
def client_ip
  remote_ip = remote_addr
  headers = Datadog::Tracing::Contrib::Rack::Header::RequestHeaderCollection.new(env)

  result = Datadog::Tracing::ClientIp.raw_ip_from_request(headers, remote_ip)

  if result.raw_ip
    ip = Datadog::Tracing::ClientIp.strip_decorations(result.raw_ip)
    return unless Datadog::Tracing::ClientIp.valid_ip?(ip)

    ip
  end
end
```

```ruby
def self.set_client_ip_tag!(span, headers: nil, remote_ip: nil)
  result = raw_ip_from_request(headers, remote_ip)

  if result.raw_ip
    ip = strip_decorations(result.raw_ip)
    return unless valid_ip?(ip)

    span.set_tag(Tracing::Metadata::Ext::HTTP::TAG_CLIENT_IP, ip)
  elsif result.multiple_ip_headers
    span.set_tag(TAG_MULTIPLE_IP_HEADERS, result.multiple_ip_headers.keys.join(','))
  end
end
```

So in the new implementation, we always remove the port and zone. In a way, the test is more in tune with how the use the code in production.